src/Nethermind/Nethermind.TxPool/NonceManager.cs

```diff
@@ -0,0 +1,99 @@
+// SPDX-FileCopyrightText: 2022 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using Nethermind.Core;
+using Nethermind.Int256;
+
+namespace Nethermind.TxPool;
+
+public class NonceManager : INonceManager
+{
+    private readonly ConcurrentDictionary<Address, AddressNonceManager> _addressNonceManagers = new();
+    private readonly IAccountStateProvider _accounts;
+
+    public NonceManager(IAccountStateProvider accounts)
+    {
+        _accounts = accounts;
+    }
+
+    public UInt256 ReserveNonce(Address address)
+    {
+        AddressNonceManager addressNonceManager =
+            _addressNonceManagers.GetOrAdd(address, _ => new AddressNonceManager());
+        return addressNonceManager.ReserveNonce(_accounts.GetAccount(address).Nonce);
+    }
+
+    public void TxAccepted(Address address)
+    {
+        if (_addressNonceManagers.TryGetValue(address, out AddressNonceManager? addressNonceManager))
+        {
+            addressNonceManager.TxAccepted();
+        }
+    }
+
+    public void TxRejected(Address address)
+    {
+        if (_addressNonceManagers.TryGetValue(address, out AddressNonceManager? addressNonceManager))
+        {
+            addressNonceManager.TxRejected();
+        }
+    }
+
+    public void TxWithNonceReceived(Address address, UInt256 nonce)
+    {
+        AddressNonceManager addressNonceManager =
+            _addressNonceManagers.GetOrAdd(address, _ => new AddressNonceManager());
+        addressNonceManager.TxWithNonceReceived(nonce);
+    }
+
+    private class AddressNonceManager
+    {
+        private HashSet<UInt256> _usedNonces = new();
+        private UInt256 _reservedNonce;
+        private UInt256 _currentNonce;
+        private UInt256 _previousAccountNonce;
+        private readonly Mutex _mutex = new();
+
+        public UInt256 ReserveNonce(UInt256 accountNonce)
+        {
+            _mutex.WaitOne();
+
+            ReleaseNonces(accountNonce);
+            _currentNonce = UInt256.Max(_currentNonce, accountNonce);
+            _reservedNonce = _currentNonce;
+            return _currentNonce;
+        }
+
+        public void TxAccepted()
+        {
+            _usedNonces.Add(_reservedNonce);
+            while (_usedNonces.Contains(_currentNonce))
+            {
+                _currentNonce++;
+            }
```

Comment:
what if `_reservedNonce` > `_currentNonce`?

Reply:
It's totally fine, we might receive tx with high nonce. This nonce will be marked as used. For example:
1. sendTx(nonce=2) => accepted
2. sendTx(nonce=null) => nonce = 0
3. sendTx(nonce=null) => nonce = 1
4. sendTx(nonce=null) => nonce = 3

Reply:
Yeah exactly my question is this scenario worth considering. Or should we just do:

    sendTx(nonce=2) => accepted
    sendTx(nonce=null) => nonce = 3
    sendTx(nonce=null) => nonce = 4
    sendTx(nonce=null) => nonce = 5

Reply:
They will never be confirmed, because of nonce gap

Reply:
Yes, but the question would be why someone would send that, maybe because smaller nonces are expected to be in TxPool already?

Reply:
We don't know in witch order transactions have been sent, we only know in what order they have been received. I think that we should assign nonces in such way that transactions have some chance of being confirmed.